src/routes/(app)/admin[[hash=admin_hash]]/discount/new/+page.server.ts

```diff
@@ -29,45 +30,90 @@ export const load = async () => {
 
 export const actions: Actions = {
 	default: async function ({ request }) {
-		const data = await request.formData();
+		const formData = await request.formData();
+		const quantityPerProductRecord: Record<string, number> = {};
+		for (const [key, value] of formData.entries()) {
+			const match = key.match(/^quantityPerProduct\[(.+?)\]$/);
+			if (match) {
+				quantityPerProductRecord[match[1]] = Number(value);
+			}
+		}
 
-		const { name, percentage, subscriptionIds, productIds, wholeCatalog, beginsAt, endsAt } = z
+		const {
+			name,
+			percentage,
+			subscriptionIds,
+			productIds,
+			wholeCatalog,
+			beginsAt,
+			endsAt,
+			quantityPerProduct,
+			mode
+		} = z
 			.object({
 				name: z.string().min(1).max(MAX_NAME_LIMIT),
 				productIds: z.string().array(),
 				subscriptionIds: z.string().array().min(1),
-				percentage: z.string().regex(/^\d+(\.\d+)?$/),
+				percentage: z
+					.string()
+					.regex(/^\d+(\.\d+)?$/)
+					.optional(),
```

Comment:
Same comment this shouldn't be optional.

You can do a follow up `z.parse` below depending on `mode`